client/views/add/add.js

```diff
@@ -9,6 +9,15 @@ function showModsError(reason) {
   currentError = Blaze.renderWithData(Template.form_error, reason, parentNode, nextNode);
 }
 
+function checkPrevMod(modBoxes) {
+  for (let i = 0; i < modBoxes.length - 1; i++) {
```

Comment:
Meteor uses [babel](https://babeljs.io/) under the covers, so you can feel free to use a more convenient form of looping, e.g. `Array.forEach`, or in this case, [`modBoxes.find`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) probably makes more sense. It'll make the code easier to read, and it'll help avoid any off-by-one errors.

Reply:
Sorry, I didn't look closely enough at the code to see what's happening here. There are a couple of issues.

First, we're only checking for duplicates when someone clicks the plus. What about when they click "Done"?
We also have a dependency on the order of the moderators. What happens if a user adds "a@b.com" first, then adds "e@f.com", then adds "a@b.com" again?

It helps to think about these things in terms of dependencies. What are all the things that need to be true for our validation to work correctly? Minimizing the number of those things makes the code easier to understand and reason about (and it usually works better). How might we minimize dependencies in this case? Well, the only things we care about are that the set of moderators is unique and that this unique number is what counts toward the limit. We're checking the limit in multiple places, so in each of those places we need to make sure we're comparing the right number. We also need to make sure the list of mods is unique when submitting it to the server. To perform both of those tasks, you can use [Sets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) to create the list of unique moderators.

Ideally, we wouldn't be checking the moderator limit or uniqueness in multiple places. Refactoring to fix that is a slightly larger job, so fixing that doesn't have to happen in this PR. We can talk about it next week. Right now, I'd just focus on using the unique set of moderators in all the places it needs to exist.

Reply:
Yes, that done button issue is a big issue as pressing only the done button also allows the user to add more than 4 moderators because the checking is only done on the plus and not the done button. Bug #15437 on Redmine discusses it.
 >What happens if a user adds "a@b.com" first, then adds "e@f.com", then adds "a@b.com" again?
 
 This works fine because the loop checks all the previous mods before adding it. Again, the check is only done by the plus button and not the done button.